src/main/java/com/solace/samples/jcsmp/snippets/HowToConfigureSessionUsingOAUTH2.java

```diff
@@ -16,55 +16,106 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import com.solacesystems.jcsmp.InvalidPropertiesException;
-import com.solacesystems.jcsmp.JCSMPException;
-import com.solacesystems.jcsmp.JCSMPFactory;
-import com.solacesystems.jcsmp.JCSMPSession;
-import com.solacesystems.jcsmp.SessionEvent;
-import com.solacesystems.jcsmp.SessionEventArgs;
-import com.solacesystems.jcsmp.SessionEventHandler;
+
+import java.time.Instant;
 
 /**
  * Demonstrates how to configure a Session using properties supporting OAuth 2.0
  * authentication and how to refresh the tokens.
  */
 public class HowToConfigureSessionUsingOAUTH2 {
-  private JCSMPSession session = null;
-
-  // remember to add log4j2.xml to your classpath
-  private static final Logger logger = LogManager.getLogger(); // log4j2, but could also use SLF4J, JCL, etc.
-
-  public void createOauth2Session(JCSMPProperties sessionProperties, String accessToken, String issuerId)
-      throws InvalidPropertiesException {
-    sessionProperties.setProperty(JCSMPProperties.AUTHENTICATION_SCHEME, JCSMPProperties.AUTHENTICATION_SCHEME_OAUTH2);
-    sessionProperties.setProperty(JCSMPProperties.OAUTH2_ACCESS_TOKEN, accessToken);
-    sessionProperties.setProperty(JCSMPProperties.OAUTH2_ISSUER_IDENTIFIER, issuerId);
-
-    // When the Access token has expired, the client will be disconnected, and
-    // will try to reconnect. Refresh the token before it reconnects.
-    session = JCSMPFactory.onlyInstance().createSession(sessionProperties, null,
-        new SessionEventHandler() {
-          @Override
-          public void handleEvent(SessionEventArgs event) {
-            if (event.getEvent() == SessionEvent.RECONNECTING) {
-              logger.info("Session Reconnecting - Refresh Access Token");
-
-              try {
-                session.setProperty(JCSMPProperties.OAUTH2_ACCESS_TOKEN, refreshAccessToken());
-              } catch (JCSMPException e) {
-                logger.error("Failed to refresh Access token");
-              }
-
-            }
-          }
-        });
-  }
-
-  private String refreshAccessToken() {
-    // TODO: Get a new access token from the identity provider
-    return "";
-  }
+    private JCSMPSession session = null;
+
+    // remember to add log4j2.xml to your classpath
+    private static final Logger logger = LogManager.getLogger(); // log4j2, but could also use SLF4J, JCL, etc.
+
+    public void createOauth2Session(JCSMPProperties sessionProperties, String accessToken, String issuerId)
+            throws InvalidPropertiesException {
+        sessionProperties.setProperty(JCSMPProperties.AUTHENTICATION_SCHEME, JCSMPProperties.AUTHENTICATION_SCHEME_OAUTH2);
+        sessionProperties.setProperty(JCSMPProperties.OAUTH2_ACCESS_TOKEN, accessToken);
+        sessionProperties.setProperty(JCSMPProperties.OAUTH2_ISSUER_IDENTIFIER, issuerId);
+
+        // When the Access token has expired, the client will be disconnected, and

```

Comment:
shouldn't we be checking to see if the token is close to being expired and refreshing it before expiry?

Reply:
Same reply as the other comment.
Once connected and authenticated from the broker, the new token will be picked only when the current token is expired and invalidated and an exception is thrown. 
As per PSGs and SEs, the recommended way is to wait for the disconnect and catch the exception and then provide a new token to try and connect.